### DISCUSSION TITLE
Batch compositing hundreds of images

### DISCUSSION THREAD

**Discussion by Dasx1** [ImageMagick/ImageMagick#7096] [Help] [answered]
### ImageMagick version

7.1.1-27

### Operating system, version and so on

Windows 10

### Description

Hi there!

I'm trying to create a gallery for a game that includes its character sprites. The sprites are split into 3 parts: body, eyes, and lips. Each 'set' of a specific pose is composed of the following: 

-1 body image (actor2_0_0.png),
![actor2_0_0](https://github.com/ImageMagick/ImageMagick/assets/159503528/1dce1934-2ef9-4b86-b69a-3b60fe5ef039)
-3 eye variations, (actor2_0_eye_1.png, actor2_0_eye_2.png, and actor2_0_eye_3),
![actor2_0_eye_1](https://github.com/ImageMagick/ImageMagick/assets/159503528/2cf9c7fa-dcf5-4d94-ab20-c0101aac9e38)
-3 lip variations (actor2_0_lip_1.png, actor2_0_lip_2.png, actor2_0_lip_3.png)
![actor2_0_lip_1](https://github.com/ImageMagick/ImageMagick/assets/159503528/25bdd777-ef2b-44f0-96f4-b1966cc3fa33)
The next set would be called actor2_1_0.png, actor2_2_0.png, and so on for all the associated image parts.

The images are in PNG format, are the same dimensions, are transparent, and are aligned properly, so a simple composite command should suffice. I would like to overlay 1 of each eye variation and 1 of each lip variation on the body image, which would be a total of 9 unique variations per pose. 

I used to copy and paste these manually, but with 100+ poses it is a huge waste of time for me to continue doing so manually. I would like to automate this process, but I'm having difficulty understanding imagemagick's syntax due to my complete lack of programming background. The closest example I've found through google is something like `FOR %i IN (actor2_0*.png) DO magick composite %i actor2_0_0.png %i_composite.png`. I don't even know what %i is, and I'm not sure where to find other similar expressions.

Any help would be much appreciated.

**Comment by snibgo** [comment, marked as answer]
For the ImageMagick command, for the examples you show, I suggest:
```
magick actor2_0_0.png actor2_0_eye_1.png -composite actor2_0_lip_1.png -composite out.png
```
Is the result correct?

Everything else is just Windows scripting. But I will help you out.

Your parts are numbered. The bodies are numbered from 0 to 100 (or whatever). For each body, we have three eyes numbered 1 to 3, and we have three lips also numbered 1 to 3. You want nine outputs per body. Is that correct so far?

I suggest one loop for the 100+ bodies. Within that loop, loop through the eyes. And within that loop, loop through the lips.

For Windows BAT, looping through numbers is done like this:
```
for /L %%A in (0,1,100) do (
  echo %%A
)
```
`%%A` is a variable. Put the above in a BAT script, and it will echo the numbers from 0 to 100. But we want a nested loop, and another loop within that one.

So we have:
```
for /L %%A in (0,1,100) do (
  echo %%A
  for /L %%B in (1,1,3) do (
    for /L %%C in (1,1,3) do (
      magick ^
        actor2_%%A_0.png actor2_%%A_eye_%%B.png -composite ^
        actor2_%%A_lip_%%C.png -composite ^
        out_%%A_%%B_%%C.png
    )
  )
)
```
EDIT: Corrected back-slash `\\` to caret `^` in the above. 

I split the `magick` command into lines for readability. This will create output files called `out_0_0_0.png`, where the three numbers are for the body, eyes and lips.

**Reply by Dasx1** [reply]
I didn't realize it was windows scripting! Thank you very much for the script and especially your explanations. It works perfectly.
